modules/exploits/freebsd/http/citrix_formssso_target_rce.rb

```diff
@@ -94,19 +124,57 @@ def initialize(info = {})
   end
 
   def check
+    # version 13.x end point
     res = send_request_cgi({
       'uri' => normalize_uri(datastore['TARGETURI'], 'logon', 'LogonPoint', 'index.html')
     })
+    return CheckCode::Unknown if res.nil?
+
+    if res.code == 200 && res.body =~ /<title class="_ctxstxt_NetscalerGateway">/
+      mytarget = get_target
+      return CheckCode::Appears("Detected #{mytarget.name}.") if mytarget
 
+      return CheckCode::Detected
+    end
+
+    # version 12.x end point
+    res = send_request_cgi({
+      'uri' => normalize_uri(datastore['TARGETURI'], 'vpn', 'index.html')
+    })
     return CheckCode::Unknown if res.nil?
 
-    return CheckCode::Safe unless res.code == 200 && res.body =~ /<title class="_ctxstxt_NetscalerGateway">/
+    if res.code == 200 && res.body =~ /Citrix Gateway/ && res.body =~ /AccessGateway\.ico/
+      mytarget = get_target
+      return CheckCode::Appears("Detected #{mytarget.name}.") if mytarget
+
+      return CheckCode::Detected
+    end
+
+    CheckCode::Safe
+  end
+
+  def get_target
+    return @detected_target if @detectection_ran
 
-    CheckCode::Detected
+    @detection_ran = true
+    res = send_request_cgi({
+      'uri' => normalize_uri(datastore['TARGETURI'], 'logon', 'fonts', 'citrix-fonts.css')
+    })
+
+    return nil unless res&.headers&.[]('Last-Modified').present?
+
+    timestamp = DateTime.parse(res.headers['Last-Modified']).to_i
```

Comment:
Just a sanity question here; I assume the the `to_i` is used here as the target server localises the time in some way? I'm guessing this was tested with having the attacker/target in different timezones and everything worked okay? 🤔

Reply:
I didn't check different timezones. I initially used the epoch to compare my results with another PoC. Everything matched but the resource was only available for v13.x. I changed mine to use a different resource that worked for both 12.x and 13.x so the timestamps are different and I just kept the integer representation.

Reply:
Sync'd up here on a call, and the header example is:

```
Last-Modified: Sat, 26 Nov 2022 12:34:13 GMT
```

And the `to_i` method comes from active support:

```
>> DateTime.parse('Sat, 26 Nov 2022 12:34:13 GMT').method(:to_i)
=> #<Method: DateTime#to_i() /Users/user/.rvm/gems/ruby-3.0.5@metasploit-framework/gems/activesupport-7.0.5/lib/active_support/core_ext/date_time/conversions.rb:84>
```

Which would technically break if active support isn't loaded, but that's a hard requirement at this point in msfconsole 😄 

```
3.0.5 :010 > DateTime.parse('Sat, 26 Nov 2022 12:34:13 GMT').to_i
(irb):10:in `<main>': undefined method `to_i' for #<DateTime: 2022-11-26T12:34:13+00:00 ((2459910j,45253s,0n),+0s,2299161j)> (NoMethodError)
Did you mean?  to_s
	from /Users/user/.rvm/gems/ruby-3.0.5@metasploit-framework/gems/irb-1.7.0/exe/irb:9:in `<top (required)>'
	from /Users/user/.rvm/gems/ruby-3.0.5@metasploit-framework/bin/irb:25:in `load'
	from /Users/user/.rvm/gems/ruby-3.0.5@metasploit-framework/bin/irb:25:in `<main>'
	from /Users/user/.rvm/gems/ruby-3.0.5@metasploit-framework/bin/ruby_executable_hooks:22:in `eval'
	from /Users/user/.rvm/gems/ruby-3.0.5@metasploit-framework/bin/ruby_executable_hooks:22:in `<main>'
```

Separately verified the parsing snippet in isolation looks to work just fine without rail's hijacking of timezone logic - will resolve :+1: